src/v2/QueryExecutor.sol

```diff
@@ -0,0 +1,204 @@
+    error InsufficientGasFee();
+
+    /// @notice Error thrown when blockhash verification fails.
+    error BlockhashMismatch();
+
+    /// @notice Error thrown when deloyed to a chain with an unknown chainId.
+    error ChainNotSupported();
+
+    /// @notice Error thrown when a non-router address calls a router-only function
+    error OnlyRouter();
+
+    modifier requireGasFee() {
+        if (msg.value < GAS_FEE) {
+            revert InsufficientGasFee();
+        }
+        _;
+    }
+
+    modifier onlyRouter() {
+        if (msg.sender != router) {
+            revert OnlyRouter();
+        }
+        _;
+    }
+
+    /// @notice Validates the query range for a storage contract.
+    /// @param startBlock The starting block number of the query range.
+    /// @param endBlock The ending block number of the query range.
+    /// @dev Reverts with appropriate errors if the query range is invalid:
+    ///      - QueryAfterCurrentBlock: If the ending block is after the current block number.
+    ///      - QueryInvalidRange: If the starting block is greater than the ending block.
+    ///      - QueryGreaterThanMaxRange: If the range (ending block - starting block) exceeds the maximum allowed range.
+    modifier validateQueryRange(uint256 startBlock, uint256 endBlock) {
+        if (SUPPORTS_L1_BLOCKDATA && endBlock > L1BlockNumber()) {
+            revert QueryAfterCurrentBlock();
+        }
+        if (startBlock > endBlock) {
+            revert QueryInvalidRange();
+        }
+        if (endBlock - startBlock + 1 > MAX_QUERY_RANGE) {
+            // NOTE: technically the max range is MAX_QUERY_RANGE-1 :facepalm:
+            revert QueryGreaterThanMaxRange();
+        }
+        _;
+    }
+
+    constructor(address _router) Ownable(msg.sender) {
+        router = _router;
+        SUPPORTS_L1_BLOCKDATA = supportsL1BlockData();
+        if (isEthereum() || isLocal()) {
+            GAS_FEE = ETH_GAS_FEE;
+        } else if (isMantle()) {
+            GAS_FEE = MANTLE_GAS_FEE;
+        } else if (isOPStack() || isCDK() || isScroll()) {
+            GAS_FEE = L2_GAS_FEE;
+        } else {
+            revert ChainNotSupported();
+        }
+    }
+
+    /// @inheritdoc IQueryExecutor
+    function request(
+        address client,
+        bytes32 queryHash,
+        bytes32[] calldata placeholders,
+        uint256 startBlock,
+        uint256 endBlock,
+        uint32 limit,
+        uint32 offset
+    )
+        public
+        payable
+        onlyRouter
+        requireGasFee
+        validateQueryRange(startBlock, endBlock)
+        returns (uint256)
+    {
+        // requestId is 12 bytes of entropy followed this 20 byte contract address
+        uint256 requestId = uint256(
+            bytes32(
+                bytes.concat(
+                    bytes12(keccak256(abi.encodePacked(++requestIDNonce))),
+                    bytes20(address(this))
+                )
+            )
+        );
```

Comment:
This is the only meaningful change between this contract and the QueryManager. Encoding the QM address in the request ID is a helpful trick that will allow the router to know where to route the response without any additional knowledge :)